ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -112,15 +145,59 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
     return right.getPartitions(kafkaTopicClient);
   }
 
+  @SuppressWarnings("UnstableApiUsage")
   @Override
   public Stream<ColumnName> resolveSelectStar(
-      final Optional<SourceName> sourceName, final boolean valueOnly
+      final Optional<SourceName> sourceName,
+      final boolean valueOnly
   ) {
-    return getSources()
-            .stream()
-            .flatMap(s -> s instanceof JoinNode ? s.getSources().stream() : Stream.of(s))
-            .filter(s -> !sourceName.isPresent() || sourceName.equals(s.getSourceName()))
-            .flatMap(s -> s.resolveSelectStar(sourceName, false));
+    final Stream<ColumnName> names = Stream.of(left, right)
+        .flatMap(s -> s instanceof JoinNode ? s.getSources().stream() : Stream.of(s))
+        .filter(s -> !sourceName.isPresent() || sourceName.equals(s.getSourceName()))
+        .flatMap(s -> s.resolveSelectStar(sourceName, false));
+
+    if (sourceName.isPresent() || joinKey.getType() != Type.SYNTHETIC || !finalJoin) {
+      return names;
+    }
+
+    final Column syntheticKey = Iterables.getOnlyElement(getSchema().key());
+
+    return Streams.concat(Stream.of(syntheticKey.name()), names);
+  }
+
+  public Expression resolveSelect(final int idx, final Expression expression) {
+    return joinKey.resolveSelect(expression, getSchema());
+  }
+
+  @SuppressWarnings("UnstableApiUsage")
+  @Override
+  void validateKeyPresent(final SourceName sinkName, final Projection projection) {
+
+    final boolean atLeastOneKey = Streams.concat(
+        joinKey.getAvailableKeys().stream(),
+        joinKey.getOriginalAvailableKeys().stream()
+    ).anyMatch(projection::containsExpression);
```

Comment:
Both are unfortunately required at the moment because of the mess around rewritten vs original expressions.

The `projection` passed in knows about rewritten column references e.g. `T_ID` , not `T.ID`. So we need to use rewritten columns returned by `joinKey.getAvailableKeys()`.  

Unfortunately, the `T_ID` won't match a `T.*` in the projection. For that to match we need the original `T.ID` - hence we need both for now.

I've got a todo locally to see if this can be avoided. Fixing it will likely mean rethinking how the rewrites are done.